### PR TITLE
[core] Add module import hook install logic

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
# Overview
This PR adds the required logic necessary for installing module import hooks.

Now integrations can register their `patch` functions to be run when the required module is imported. This means that integrations can be certain of being the first to run after the code is imported.

What this looks like for integrations (with `gevent` as the example). In `gevent/__init__.py`:

```python
def _patch(gevent):
    # install logic that will be run when module is imported

def patch():
    # install the module import hook to patch gevent on module import
    install_module_import_hook('gevent', _patch)

def unpatch():
    # uninstall logic...
    # remove import hook if it exists
    uninstall_module_import_hook('gevent')
```

# Caveats

This approach uses `wrapt`'s module import hooking system which [deletes the hooks after they are run](https://github.com/GrahamDumpleton/wrapt/blob/4bcd190457c89e993ffcfec6dad9e9969c033e9e/src/wrapt/importer.py#L127-L136). This isn't desirable for our use-case even though it's highly unlikely that modules will be reimported or deleted. Another undesired effect of using `wrapt` is that we cannot uninstall module import hooks and have to import `wrapt` privates to do so. While it is overkill for this PR to introduce our own meta-path managing it is something we _will_ have to move to long term.

### CODE REVIEW

File: tests/utils/test_install.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from ddtrace.utils.install import (
+    install_module_import_hook,
+    uninstall_module_import_hook,
+    _mark_module_patched,
+    _mark_module_unpatched,
+    module_patched,
+)
+
+
+@run_in_subprocess
```
Comment (Kyle-Verhoog) [inline tests/utils/test_install.py]: `@run_in_subprocess` 👌

**Comment by brettlangdon** [comment]
> While it is overkill for this PR to introduce our own meta-path managing it is something we will have to move to long term.

I wonder if we shouldn't just do this first? It feels like we are setting ourselves up for failure here.

We might not see an edge case where this is a problem right now, but a customer will definitely find one for us... especially if we already can see this being an issue in the future.

File: tests/utils/test_install.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+import mock
+import sys
+import types
+
+from tests.subprocesstest import SubprocessTestCase, run_in_subprocess
+
+from ddtrace.utils.install import (
```
Comment (brettlangdon) [inline tests/utils/test_install.py]: We should probably have this as `ddtrace.utils.import_hook` to be more clear, since `install` could be anything, and makes me think about `pip install` first before module hooks.

File: tests/utils/test_install.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    install_module_import_hook,
+    uninstall_module_import_hook,
+    _mark_module_patched,
```
Comment (brettlangdon) [inline tests/utils/test_install.py]: if we are going to be testing these, should they remain "_private"?

Comment (brettlangdon) [reply tests/utils/test_install.py]: eh, probably, but feels weird is all

File: tests/utils/test_install.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            def patch_fn(self):
+                return 2
+            mod.A.fn = patch_fn
+        install_module_import_hook('tests.utils.test_module', hook)
+
+        import tests.utils.test_module  # noqa
+        self.assertEqual(tests.utils.test_module.A().fn(), 2)
+
+        # TODO: the assert below should be
+        # >>> assert tests.utils.test_module.A().fn() == 2
```
Comment (brettlangdon) [inline tests/utils/test_install.py]: this should still pass?

If we reloaded the module then it would fail.

Comment (brettlangdon) [reply tests/utils/test_install.py]: oh, are you referring to L68?

**Comment by Kyle-Verhoog** [comment]
I don't think we're quite setting ourselves up for failure since this is quite an edge-case (most modules break or have undefined behaviour when reimported/deleted and imported again). Plus there is the fact that our current patching does not account for reimports so the functionality is the same from the user's perspective (they would have to call `ddtrace.patch(module=True)` again after deleting the module.

With that said, it's something we're going to change out eventually so it's worth considering not introducing it in the first place. I will investigate bringing in the core pieces of `wrapt` or [`importhook`](https://github.com/brettlangdon/importhook) with the modification that give us what we want from the start.

File: tests/utils/test_install.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        # remove the module from sys.modules to force a reimport
+        del sys.modules['tests.utils.test_module']
+        import tests.utils.test_module  # noqa
+        self.assertEqual(tests.utils.test_module.A().fn(), 1)
+
+    def test_install_module_import_hook_already_imported(self):
+        """
+        Test that import hooks are fired when the module is already imported.
+        """
+        module = types.ModuleType('somewhere')
```
Comment (brettlangdon) [inline tests/utils/test_install.py]: Why not `import tests.utils.test_module  # noqa`?

**Comment by brettlangdon** [comment]
Fair enough. My only concern here is if we know what we probably need, if we hold off on it, will we ever implement it?

File: tests/utils/test_install.py (hunk 5 above)
Comment (brettlangdon) [reply tests/utils/test_install.py]: It might be better to test with a real module, just in case there is a weird edge case with `sys.modules['mod'] = mod`... maybe?

File: ddtrace/utils/install.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
+import sys
+
+from wrapt.importer import (
+    register_post_import_hook,
+    _post_import_hooks,
```
Comment (brettlangdon) [inline ddtrace/utils/install.py]: I think I actually know the answer, but what versions of `wrapt` are all of these functions available for?

Since we don't lock `wrapt` to a specific version we might shoot ourselves in the foot if a customer either has an old version pinned, or a new version changes these.

File: ddtrace/utils/install.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+def uninstall_module_import_hook(modulename):
+    """Uninstalls a module import hook for a module given its name.
+
+    Note: any patching done on the module will still apply, this function
+    simply removed/disables the import hook.
+    """
+    if modulename in sys.modules:
+        module = sys.modules[modulename]
+        _mark_module_unpatched(module)
```
Comment (brettlangdon) [inline ddtrace/utils/install.py]: Not sure we want this?

What if we install, patch, then uninstall? The module should still be marked as patched?

File: tests/utils/test_install.py (hunk 5 above)
Comment (Kyle-Verhoog) [reply tests/utils/test_install.py]: Good point. This is remnant from pre `@run_in_subprocess`.. dark days. It can be updated to use the `test_module`.

File: tests/utils/test_install.py (hunk 4 above)
Comment (brettlangdon) [reply tests/utils/test_install.py]: Might be better to locate the comment closer to the line it is referring to.

File: ddtrace/utils/install.py (hunk 6 above)
Comment (Kyle-Verhoog) [reply ddtrace/utils/install.py]: Yeah with this approach we'd have to lock wrapt.. which we should be doing anyways.

Comment (brettlangdon) [reply ddtrace/utils/install.py]: I would love your review on #751, as it tried to vendor `six`.

File: ddtrace/utils/install.py (hunk 7 above)
Comment (Kyle-Verhoog) [reply ddtrace/utils/install.py]: Yeah it's not ideal but `uninstall_module_import_hook` assumes that the caller is going to perform the unpatching logic and then call `uninstall_module_import_hook` and hence we mark it "unpatched". Previously I had exposed the `mark_module_unpatched` and then the convention would be for the integration to be in charge of marking it unpatched. Neither of these solutions seem particularly good.

Comment (Kyle-Verhoog) [reply ddtrace/utils/install.py]: maybe `uninstall_module_import_hook_and_mark_unpatched` is a better more verbose solution?

Comment (brettlangdon) [reply ddtrace/utils/install.py]: If we are moving away from "unpatching" do we need to mark it as unpatched at all?

We should have `patch()` -> setup hooks, `unpatch()` -> disable hooks  ?

File: ddtrace/utils/hook.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+from wrapt.decorators import synchronized
+
+
+PY2 = sys.version_info[0] == 2
+PY3 = sys.version_info[0] == 3
+
+if PY3:
+    string_types = str,
+else:
+    string_types = basestring,
```
Comment (brettlangdon) [inline ddtrace/utils/hook.py]: we have these as apart of `ddtrace.compat`

File: ddtrace/utils/hook.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
+import sys
```
Comment (brettlangdon) [inline ddtrace/utils/hook.py]: If this is copy/pasta from `wrapt`, then we should have a comment here on that. GitHub link to tag/commit hash version of the file we copied, along with a description of changes we have made (maybe inline comments about what we have changed?)

File: tests/utils/test_import_hook.py [hunk 10]
```diff
@@ -57,24 +57,17 @@ def patch_fn(self):
         import tests.utils.test_module  # noqa
         self.assertEqual(tests.utils.test_module.A().fn(), 2)
 
-        # TODO: the assert below should be
```
Comment (brettlangdon) [inline tests/utils/test_import_hook.py]: :+1: 🔥

File: ddtrace/utils/install.py (hunk 7 above)
Comment (Kyle-Verhoog) [reply ddtrace/utils/install.py]: _Eventually_ yes I think. It's just tricky for what we do in the meantime.

File: ddtrace/utils/hook.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    were registered against the target module will be invoked. If an
+    exception is raised in any of the post import hooks, that will cause
+    the import of the target module to fail.
+    """
+    name = getattr(module, '__name__', None)
+    hooks = _post_import_hooks.get(name, None)
+
+    if hooks:
+        for hook in hooks:
+            hook(module)
```
Comment (brettlangdon) [inline ddtrace/utils/hook.py]: We should probably add a `try/except` here, from the sound of the docstring above it is intentional to raise the exception, but we should catch and log it.

Comment (Kyle-Verhoog) [reply ddtrace/utils/hook.py]: Great idea.

File: ddtrace/utils/hook.py [hunk 12]
```diff
@@ -1,16 +1,27 @@
...
+"""
 import sys
 import threading
 
 from wrapt.decorators import synchronized
 
+from ddtrace.compat import string_type
+
 
 PY2 = sys.version_info[0] == 2
```
[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply ddtrace/utils/hook.py]: I'll use `PY2` I think it makes sense to add `PY3` to `ddtrace.compat` as well.

File: tests/utils/test_import_hook.py [hunk 13]
```diff
@@ -91,3 +91,21 @@ def test_install_uninstall_install(self):
...
+            raise Exception('module hook failed')
+
+        with mock.patch('ddtrace.utils.hook.log') as log_mock:
+            install_module_import_hook('tests.utils.test_module', err_hook)
+            import tests.utils.test_module  # noqa
+            calls = [
+                mock.call('failed to call hook for module "tests.utils.test_module": module hook failed'),
+            ]
+            import sys
+            sys.stderr.write(str(log_mock.warn.mock_calls))
```
Comment (brettlangdon) [inline tests/utils/test_import_hook.py]: ?

File: tests/utils/test_import_hook.py [hunk 14]
```diff
@@ -91,3 +91,21 @@ def test_install_uninstall_install(self):
...
+            log_mock.warn.assert_has_calls(calls, any_order=True)
```
Comment (brettlangdon) [inline tests/utils/test_import_hook.py]: `log_mock.warn.assert_called_once()`?

File: tests/utils/test_import_hook.py (hunk 13 above)
Comment (Kyle-Verhoog) [reply tests/utils/test_import_hook.py]: ah crap

File: tests/utils/test_import_hook.py (hunk 14 above)
Comment (Kyle-Verhoog) [reply tests/utils/test_import_hook.py]: should be equivalent right?

Comment (brettlangdon) [reply tests/utils/test_import_hook.py]: yeah, should be

**Comment by brettlangdon** [comment]
@Kyle-Verhoog I'll take it from here
